### ISSUE TITLE
Release Agda 2.6.1 release candidate 1

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#4212]
The time for the next release is coming close. We will not make our (optimistic) 1 December deadline, but it would be nice if we could release 2.6.1 before Christmas (or at least have a RC).

To avoid the delays and crunch period we had with the release of 2.6.0, I want to propose a policy to decide which issues must be fixed and which ones can be bumped. Let's define a **critical issue** as one in the following categories:

- Regressions in the current master branch (label `regression on master`): #3960, #4005, #4040, #4149, #4169
- Regressions in the latest major release (label `regression in 2.6.0`): #3054, #3538, #4067
- Bugs in unreleased features (labels `type: bug` and `not-in-changelog`): #3995, #4137, #4169, #4172, #4174, #4195, #4200

All other issues I consider **non-critical**.

I would like to ask everyone who has at least one open issue with the 2.6.1 milestone (@UlfNorell @andreasabel @asr @gallais @Saizan @ice1000 @xekoukou @L-TChen @hborum @mortberg @fredrikNordvallForsberg @Blaisorblade) to go over the list of 2.6.1 issues (https://github.com/agda/agda/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.6.1) assigned to you and do the following:

- If it is a *critical issue* (as defined above): decide whether you realistically will be able to work on it before Christmas, or else **post a comment asking for help** (and possibly also unassign yourself).
- If it is a *non-critical issue*: choose one of the options above, *or* if you plan to work on it but not before the release, change the milestone to 2.6.2.

If you have the time, it would also be greatly appreciated if you could do a pass over the unassigned issues with 2.6.1 milestone (https://github.com/agda/agda/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.6.1+no%3Aassignee) to decide if there are any you want to see fixed, and either assign yourself or post a comment asking if someone else can take it.

To avoid delays, please try to do these things before **December the 1st**. After that, I will start mercilessly bumping non-critical issues with no owner to 2.6.2.

Also please let me know if there are other categories of critical issues that I have missed, or if you have other comments about this proposed policy.

**Comment by guillaumebrunerie** [comment]
Hi Jesper, please take a look at my latest comment at https://github.com/agda/agda/issues/4122#issuecomment-551844309 (maybe I should have opened a new issue for it). It makes using Prop much harder than it was before (and is a regression on master).
Thanks!

**Comment by Blaisorblade** [comment]
Hi Jesper, see my summary on this issue: https://github.com/agda/agda/issues/2009#issuecomment-555460908.

**Comment by ice1000** [comment]
I'll be closing the only issue (#4183) assigned me via #4213, which is a finished and waiting-for-CI PR.

**Comment by asr** [comment]
We should add to the list of critical issues the support of new versions in the tool-chain (GHC, cabal-install, stack, alex, happy, etc). I filed #4216.

**Comment by jespercockx** [comment]
> We should add to the list of critical issues the support of new versions in the tool-chain (GHC, cabal-install, stack, alex, happy, etc). I filed #4216.

Should we have a label for classifying these issues perhaps? That would make it easier to tell which issues should definitely not be bumped.

**Comment by nad** [comment]
>     * Bugs in unreleased features (labels `type: bug` and `not-in-changelog`): (no current issues)

I added the label `not-in-changelog` to some open issues related to `@0` (#3855, #4137, #4169, #4172, #4174, #4195, and #4200).

**Comment by jespercockx** [comment]
> > ```
> > * Bugs in unreleased features (labels `type: bug` and `not-in-changelog`): (no current issues)
> > ```
> 
> I added the label `not-in-changelog` to some open issues related to `@0` (#3855, #4137, #4169, #4172, #4174, #4195, and #4200).

Ok, I've added them to the OP.

**Comment by nad** [comment]
Why not #3855? Oh, because it is an enhancement. Well, I think it makes sense to document `@0` before releasing it. (I guess that technically it was released with Agda 2.6.0, but it did not work, and it was not mentioned in the changelog.)

I removed the `not-in-changelog` label from this issue, because I guess it makes sense to include it in the changelog.

**Comment by nad** [comment]
A new regression: #4253.

(Instead of manually listing things to fix in this ticket, wouldn't it be better to change the milestones of the things that we do not plan to fix before the release?)

**Comment by jespercockx** [comment]
> (Instead of manually listing things to fix in this ticket, wouldn't it be better to change the milestones of the things that we do not plan to fix before the release?)

I am planning to do this very soon, but as noted in the original post I wanted to give people the chance to pick up issues they want to see fixed before changing all milestones. Maybe it would make more sense to change all the milestones *first* and then change it back if necessary (something to think about for next release).

**Comment by nad** [comment]
Another new regression: #4254.

**Comment by jespercockx** [comment]
Thank you to everyone for your hard work so far. I bumped all the remaining unassigned non-critical issues to 2.6.2 (except for a few that I assigned to myself). This leaves us with 40 remaining open issues, of which:

- 10 are unassigned critical issues
- 12 are assigned to myself
- 4 are assigned to @andreasabel 
- 3 are assigned to @UlfNorell 
- 2 are assigned to @asr 
- 2 are assigned to @gallais 
- 1 is assigned to each of @Saizan @ice1000 @erydo @L-TChen @ggreif @mortberg 

Let's continue the hard work so we can get our Christmas release of Agda!

**Comment by L-TChen** [comment]
The issue #4072 is blocked by the license issue, and I don’t really know how to progress it further. Otherwise everything else is ready. It would be nice if @nad could kindly give me a hand. 

I hope it can make contributing to Agda or the standard library more accessible to more people.

**Comment by nad** [comment]
That code has presumably been released under the same licence as Agda. (At least that was my intention, I have not asked other contributors.) However, I am still sceptical towards moving `fix-agda-whitespace` to a separate repository, as documented in the ticket.

**Comment by L-TChen** [comment]
Okay, thanks. Now I put it into the ice box, so I don’t have any open issue assigned to me for now.

**Comment by asr** [comment]
I need to catch up with all the recent changes to the CI as preparation for the release. I'll create the `release-2.6.1` branch from the current `master` and make the necessary changes to the CI (e.g. test with all the supported GHC versions) for releasing.

**Comment by asr** [comment]
Travis is reporting the following [error](https://www.travis-ci.org/agda/agda/jobs/624858521) in the `release-2.6.1` branch: 

```
/home/travis/.travis/functions: line 109: eval: - : invalid option
eval: usage: eval [arg ...]
```

Any idea what is wrong?

**Comment by L-TChen** [comment]
My bad. Please remove the bar `|` on https://github.com/agda/agda/blob/29790fab3c71d1c92a09b4a589b02360c7ebaf20/.travis.yml#L190

**Comment by asr** [comment]
Done. Thanks!

**Comment by asr** [comment]
@L-TChen, there are more [errors](https://www.travis-ci.org/agda/agda/jobs/624883720) related to the `.travis.yml` file reported by Travis on the `release-2.6.1` branch. Since we don't merge the release branches into `master`, you should fix the errors in `master` and merge `master` into `release-2.6.1`.

**Comment by asr** [comment]
I added the [list](https://github.com/agda/agda/blob/master/CHANGELOG.md#list-of-all-closed-issues) of *current* closed issues to the `CHANGELOG`. Tell me if you find a missing or wrong listed issue.

**Comment by ice1000** [comment]
The CHANGELOG.md is way too big. I think it's good to split it (version-per-file or some-versions-per-file or something)

**Comment by ice1000** [comment]
@asr do you have a script for generating the "closed issues" section?

**Comment by asr** [comment]
> @asr do you have a script for generating the "closed issues" section?

Yes. I use the `closed-issues-by-milestone` program in `src/release-tools/` written by @andreasabel. For using the program, see [here](https://github.com/agda/agda/blob/master/HACKING.md#closed-issues-by-milestone-program). To know the closed issues reported in the CHANGELOG, see [here](https://github.com/agda/agda/blob/master/HACKING.md#closed-issues-reported-in-the-changelog).

**Comment by asr** [comment]
I realised the issues labelled `regression in master` should not be included in the CHANGELOG. I'll fix it.

If you want issues with label `XYZ` don't show up in the CHANGELOG, please added `XYZ` to the list https://github.com/agda/agda/blob/330dd010cf0f5c92bdcb8302d94b547e4dc0ea32/src/release-tools/closed-issues-for-milestone/Main.hs#L66-#L85.

**Comment by asr** [comment]
I updated the [list](https://github.com/agda/agda/blob/master/CHANGELOG.md#list-of-all-closed-issues) of closed issues in the `CHANGELOG`.

**Comment by nad** [comment]
I did not study the list carefully, but I noticed that #3856 is included. This is not a bug fix, and it is documented elsewhere in the changelog. Should issues like this be included? I suggest that we only include bug fixes for released features.

**Comment by asr** [comment]
If every new feature were documented in *both* the CHANGELOG and the user manual, we could use the section 'List of all closed issues' for listing only bug fixes. Unfortunately, this is not the case.

BTW, some times an entry in the CHANGELOG is more complete than the correspondent entry in the user manual (sometimes even there is no entry in the user manual). I think it should be the other way around, that is, a small entry in the CHANGELOG linking to a more complete description in the user manual.

**Comment by nad** [comment]
> Unfortunately, this is not the case.

If some new feature is not mentioned in the changelog, then this should be fixed, and adding a link to the bug tracker in the changelog does not make the user manual better.

> I think it should be the other way around

I do not agree completely. In some cases I include some information in the changelog to explain, say, subtle changes to some command, whereas I do not include this information in the user manual because I feel that these subtle changes are irrelevant to people who have not used the command before.

**Comment by asr** [comment]
I'm getting this [error](https://github.com/agda/agda/commit/064095e14042bdf64c7d7c97c2869f63f5f1f8f6/checks?check_suite_id=364600233) in the `cabal build` workflow. I suppose that the error is due to [this](302ffed08a023b29eb898a0ef3968aca452f6d5a) change which was required by `cabal install 3.0.0.0`. I guess we need to modify `actions/checkout@v1`  for also cloning the standard library.

Any idea?

CC' @L-TChen.

**Comment by asr** [comment]
> If some new feature is not mentioned in the changelog, then this should be fixed.

I agree. How we know if there are new features not mentioned in the CHANGELOG?

**Comment by L-TChen** [comment]
> I'm getting this [error](https://github.com/agda/agda/commit/064095e14042bdf64c7d7c97c2869f63f5f1f8f6/checks?check_suite_id=364600233) in the `cabal build` workflow. I suppose that the error is due to [this](302ffed08a023b29eb898a0ef3968aca452f6d5a) change which was required by `cabal install 3.0.0.0`. I guess we need to modify `actions/checkout@v1` for also cloning the standard library.

I am not able to see the link (which is broken), but I guess it is because submodules are not checked out by default. 

You can check https://github.com/actions/checkout to see how to check out submodules or other repository in that step.

**Comment by asr** [comment]
> I am not able to see the link (which is broken).

The link works for me.

**Comment by asr** [comment]
@L-TChen, which is the purpose of the `cabal build` workflow?

BTW, the time reduction in the CI is great. Thank you for your hard work!

**Comment by L-TChen** [comment]
The idea is to make sure `cabal new-build` compilation works. Probably we can document how to install Agda via cabal and stack based on the CI scripts ...? 

The old-style is likely to deprecate in the following years, so I don’t want to prompt this installation method. But, if you feel necessary, I can also add testing for `cabal v1-build` for earlier GHC versions.

**Comment by asr** [comment]
> But, if you feel necessary, I can also add testing for `cabal v1-build` for earlier GHC versions.

I think it is not necessary.

**Comment by asr** [comment]
@L-TChen, we have various Github workflows. Which workflows are run/called on the `issue-xyz` branches? I'm asking because in the branch `issue-4305` the `cabal build` was not run/called.

**Comment by L-TChen** [comment]
Currently, GitHub workflows are run on all branches. However, they are run only if relevant change is made. For example, `Document` will be run if anything is changed under the directory `doc`. 

For `cabal`, this workflow is triggered if anything under `src/` is changed or `Agda.cabal` is changed. Please add files or directories to the list if you have other relevant files that should trigger the workflow.

**Comment by asr** [comment]
I changed the file `.github/workflows/cabal.yml`. Should not the `cabal build` workflow be *automatically* triggered in this case?

> Please add files or directories to the list if you have other relevant files that should trigger the workflow.

Which list?

**Comment by L-TChen** [comment]
https://github.com/agda/agda/blob/064095e14042bdf64c7d7c97c2869f63f5f1f8f6/.github/workflows/cabal.yml#L3-L11

You are right. These configurations themselves should be added.

**Comment by asr** [comment]
Please add them.

**Comment by ice1000** [comment]
> I changed the file .github/workflows/cabal.yml. Should not the cabal build workflow be automatically triggered in this case?

Speaking of stack.yml they'll be triggered if `.github/` things changed

**Comment by L-TChen** [comment]
The trigger list has been updated by 00180cdca1a189ffa68810365ed2f60431bc352b. @asr, I also include your proposed change, i.e. checking out submodules, to the commit.

**Comment by JacquesCarette** [comment]
I am trying to get a version of 2.6.1 compiling locally, so that I can test `agda-categories` with it, to ensure no additional regressions sneak in.  But I'm having a difficult time with getting it 'ready', mostly because it seems that the `HACKING.md` instructions are not fully in-synch with 2.6,1.

Just to get started: should I be using the `cabal` v1 route, or the `stack` route? Does it matter? [I've experimented, things go weird, but I'd rather use the recommended one, and then report problems with it.]

**Comment by asr** [comment]
> Just to get started: should I be using the `cabal` v1 route, or the `stack` route? Does it matter?

It does not matter. Both routes are working.

**Comment by L-TChen** [comment]
@JacquesCarette, `stack` and `cabal` v1 or even v2 (with GHC 8.8.1) all are tested and should be working. If you’d like to, can you report the problem you have in another ticket?

**Comment by JacquesCarette** [comment]
Ok. I'll work on it systematically, in the hopes of being able to file reproducible bugs (in separate tickets). I'll have time this afternoon and tomorrow.

**Comment by asr** [comment]
@JacquesCarette, thanks for thinking on test `agda-categories` with the upcoming 2.6.1. In fact, I found a problem but I'll wait for you report it using a minimal example if possible.

**Comment by asr** [comment]
I'll be without Internet access from 25th of December to January 1st. I think the `release-2.6.1` branch is in a good shape for preparing and releasing a release candidate today (mentioning the current open issues) or we can wait until the first weeks of January for releasing the release candidate hoping without knowing regressions. What do you think?

I suggest to release the release candidate today. In this way, we could release a second release candidate or the final version in the first weeks of January.

**Comment by jespercockx** [comment]
I'm in favour of bringing out a release candidate today! It will make for a nice Christmas present.

**Comment by JacquesCarette** [comment]
I've found at least 2 regressions - will post some reports as soon as I've managed to make reasonably minimal examples.

**Comment by asr** [comment]
I'm starting the preparation of the release candidate from the current master (40034c56a90f847f3299eff5d1c10290b00ea91c), so the `release-2.6.1` branch is frozen.

**Comment by asr** [comment]
I couldn't prepare the release candidate for 2.6.1 using `cabal-install 3.0.0.0` due to https://github.com/haskell/cabal/issues/5813.

**Comment by JacquesCarette** [comment]
@L-TChen I can definitely report that I was able to get going, with a minimum of fuss, with `stack`. The overall instructions could be improved [which I hope to attempt to do myself], but it worked.

The good news is that I was able to apply a workaround to both regressions, so that agda-categories currently works with both 2.6.0.1 and 2.6.1 (and stdlib-1.2).

**Comment by asr** [comment]
I uploaded to Hackage the release candidate [Agda-2.6.0.1.20191219](https://hackage.haskell.org/package/Agda-2.6.0.1.20191219/candidate) and I tested its installation on Ubuntu with

| GHC | cabal-install |
| --- | ------------- |
| 8.8.1 | 3.0.0.0 |
| 8.6.5 | 2.4.1.0 |
| 8.4.4 | 2.2.0.0 |
| 8.2.2 | 2.0.0.1 |
| 8.0.2 | 1.24.0.2 |

Could someone test the installation on macOS and Windows using the following instructions:

```bash
$ cabal get https://hackage.haskell.org/package/Agda-2.6.0.1.20191219/candidate/Agda-2.6.0.1.20191219.tar.gz
$ cd Agda-2.6.0.1.20191219
$ cabal install
```

**Comment by ggreif** [comment]
The middle command should be
```
cd Agda-2.6.0.1.20191219
```
Installing gave me
```
$ cabal install
cabal: Unknown target '.'.
There is no component '.'.
The project has no package directory '.'.
```
Not sure what this is.
I am on MacOS X High Sierra,
```
[nix-shell:~/agda/Agda-2.6.0.1.20191219]$ cabal --version
cabal-install version 3.0.0.0
compiled using version 3.0.0.0 of the Cabal library

[nix-shell:~/agda/Agda-2.6.0.1.20191219]$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.6.5
```

When I try
```
$ cabal v2-install agda agda-mode --installdir=/Users/ggreif/.cabal/bin
Wrote tarball sdist to
/Users/ggreif/agda/dist-newstyle/sdist/Agda-2.6.1.tar.gz
Wrote tarball sdist to
/Users/ggreif/agda/dist-newstyle/sdist/fix-agda-whitespace-0.0.4.tar.gz
Resolving dependencies...
Up to date
Symlinking 'agda'
Symlinking 'agda-mode'
```
it seems like all build artefacts are already built... Strange.

When I do `cabal v2-build agda agda-mode` it starts to build. It chugs along...
And it succeeds building!

**Comment by asr** [comment]
> The middle command should be
> ```
> cd Agda-2.6.0.1.20191219
>```

Oops! Fixed. Thanks!

**Comment by asr** [comment]
> Installing gave me
> ```
> $ cabal install
> cabal: Unknown target '.'.
> There is no component '.'.
> The project has no package directory '.'.
>```
>Not sure what this is.
>I am on MacOS X High Sierra,
>```
>[nix-shell:~/agda/Agda-2.6.0.1.20191219]$ 

I could reproduce the error by installing `Agda-2.6.0.1.20191219` *inside* the directory where I have the upstream version. I guess the error is due to the `cabal.project` file in this directory.

Please move `Agda-2.6.0.1.20191219` outside the `~/agda` directory and try again.

**Comment by ggreif** [comment]
> Please move `Agda-2.6.0.1.20191219` outside the `~/agda` directory and try again.

Yep, that seems to work:
```
[nix-shell:~/Agda-2.6.0.1.20191219]$ cabal v2-install agda agda-mode --installdir=/Users/ggreif/.cabal/bin -j6
Wrote tarball sdist to
/Users/ggreif/Agda-2.6.0.1.20191219/dist-newstyle/sdist/Agda-2.6.0.1.20191219.tar.gz
Resolving dependencies...
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - Agda-2.6.0.1.20191219 (exe:agda, exe:agda-mode) (requires build)
Starting     Agda-2.6.0.1.20191219 (all, legacy fallback)
Building     Agda-2.6.0.1.20191219 (all, legacy fallback)
Installing   Agda-2.6.0.1.20191219 (all, legacy fallback)
Completed    Agda-2.6.0.1.20191219 (all, legacy fallback)
Symlinking 'agda'
cabal: Path '/Users/ggreif/.cabal/bin/agda' already exists. Use
--overwrite-policy=always to overwrite.


[nix-shell:~/Agda-2.6.0.1.20191219]$ cabal v2-install agda agda-mode --installdir=/Users/ggreif/.cabal/bin -j6 --overwrite-policy=always
Wrote tarball sdist to
/Users/ggreif/Agda-2.6.0.1.20191219/dist-newstyle/sdist/Agda-2.6.0.1.20191219.tar.gz
Resolving dependencies...
Up to date
Symlinking 'agda'
Symlinking 'agda-mode'
```

**Comment by asr** [comment]
Thanks for testing. In your case I guess something like 

```
$ cabal install -j6
```

should be enough.

**Comment by asr** [comment]
Could someone test the release candidate on Windows?

**Comment by asr** [comment]
> I did not study the list carefully, but I noticed that #3856 is included. This is not a bug fix, and it is documented elsewhere in the changelog. Should issues like this be included? I suggest that we only include bug fixes for released features.

I'll added the label `documented-in-changelog` for issues which are already documented in the CHANGELOG. These issues will not include in the list of closed issues.

**Comment by asr** [comment]
I updated the [list](https://github.com/agda/agda/blob/master/CHANGELOG.md#other-closed-issues) of other closed issues in the `CHANGELOG`. If you closed an issue which wasn't a bug but an enhancement, please added one entry to the CHANGELOG. In this way, we can change the title 'Other closed issues' to 'Other fixed bugs'.

**Comment by asr** [comment]
Done. Released the release candidate.
